Add empty card placeholder

emiham opened the pull request:
If you have vocab cards and your first field is the vocab or its translation, you aren't able to create cards, only modify existing ones.

After this PR you can add a placeholder text to the first field of a card, e.g. "TODO", that you can later search for and edit.

Sorry for the README edits, they snuck in without me noticing, but I figure there's no harm in including them.

tatsumoto-ren commented:
This is actually mentioned in the README.

> Sentence field should be first in the note type settings. Otherwise, Anki won't allow mpvacious to add new notes.

So you just need to get a note type where the sentence field is set as the first field. [Like this](https://ankiweb.net/shared/info/1557722832).

Most of the time it's a bad idea to have vocab cards. An example sentence should be placed [on the front](https://tatsumoto.neocities.org/blog/discussing-various-card-templates). If you're mining cards from anime, for example, it makes no sense to omit the sentence. 

If you have a custom note type where for some reason you can't change the order of the fields then it is worth considering this PR, but it would be great if you provided more info about why that is the case.

emiham commented:
> Most of the time it's a bad idea to have vocab cards.

I don't disagree, but I don't think people shouldn't be allowed to create them. 

> If you have a custom note type where for some reason you can't change the order of the fields then it is worth considering this PR, but it would be great if you provided more info about why that is the case.

If you have the vocab field as your first field this lets you use Anki's built in duplicate detection to see if you have a card for that word already. At least in the past this was was Yomitan or one of its predecessors used for duplicate detection, not sure if it has changed.

tatsumoto-ren commented:
The successor of the now abandoned Yomichan is [Rikaitan](https://rikaitan.github.io/).

emiham commented:
> The successor of the now abandoned Yomichan is [Rikaitan](https://rikaitan.github.io/).

The "official" successor is Yomitan. I didn't actually know about Rikaitan, so I'm curious about the differences. I'll make a separate discussion topic for that though. 

> There shouldn't probably even be a setting for this. just check if sentence_field != first field and if so then set note[first_field] to "empty". And update the README accordingly.

That makes sense, I'll remove the setting.

tatsumoto-ren commented:
Rikaitan is the official successor to Yomichan. Please refrain from mentioning the other fork of Yomichan anywhere, it is an attempt to hijack the (former) Yomichan project by a malicious third-party. No one should know that it exists. Compared to Rikaitan, it offers no advantages. The fact that Alex (foosoft) advertised it should not carry any weight, as this decision discredits him.

emiham commented:
I removed the config option and updated the README. Feel free to change the wording. 

Also let me know if you want me to squash the commits or anything.

tatsumoto-ren commented on subs2srs.lua:
```diff
@@ -388,6 +389,9 @@ local function export_to_anki(gui)
     audio.run_async()
 
     local note_fields = construct_note_fields(sub['text'], sub['secondary'], snapshot.filename, audio.filename)
+    if not h.is_empty(first_field) then
+        note_fields[first_field] = "[empty]"
+    end
```
When I run this on my computer, my cards don't contain sentences anymore. There's "empty" instead. I think the condition should be like this.

``` lua
if not h.is_empty(first_field) and h.is_empty(note_fields[first_field]) then
    note_fields[first_field] = "[empty]"
end
```

emiham replied:
Nice catch. I had that line in there but it disappeared at some point.

tatsumoto-ren commented:
Since ` ankiconnect.get_first_field(config.model_name)` is being called before every card creation, would it be better to get rid of the global variable `first_field` and instead move the function call where the result is used?

emiham commented:
It won't be called if you haven't enabled `reload_config_before_card_creation`. So then you get one less API call, which probably doesn't matter, but overhead is overhead. 

Honestly though I can see the argument for just calling it every time. I figured if you want to reload the config there is a greater chance that you've changed the order of the fields (not that this is something that happens often at all). In actuality there's no reason to assume those events are actually correlated though, so the safest thing would be to accept the tiny overhead and just check every time.

tatsumoto-ren commented:
If we somehow could get rid of the global variable `first_field`, that would be great. Maybe just putting it inside `config` is enough though.

emiham commented:
A config option would work, but I think that adds unnecessary friction for new users. It's also strictly speaking not really a configuration option, it's just an chore to make sure the setting matches your note structure.
